### PR TITLE
Feature: Improve Kernel Decorator

### PR DESCRIPTION
**Pull request by smazouz42**
This pull request addresses issue #68 by changing the implantation of kernel decorate, so the function runs multiple times depending on the number of blocks and the number of threads for each block

### CODE REVIEW

File: pyccel/decorators.py [hunk 1]
```diff
@@ -130,6 +130,40 @@ def kernel(f):
...
+            self._thread_idx = thread_idx
+
+        def threadIdx(self, dim):
+            """
+            Get the thread index.
+
+            Get the thread index.
+            """
+            return self._thread_idx
+
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/decorators.py:150]: /bot accept 
this will be tested https://github.com/pyccel/pyccel-cuda/pull/66

File: pyccel/decorators.py [hunk 2]
```diff
@@ -130,6 +130,40 @@
...
+        def blockIdx(self, dim):
+            """
+            Get the block index.
+
+            Get the block index.
+            """
+            return self._block_idx
+
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/decorators.py:158]: /bot accept 
this will be tested https://github.com/pyccel/pyccel-cuda/pull/66

File: pyccel/decorators.py [hunk 3]
```diff
@@ -130,6 +130,40 @@
...
+        def blockDim(self, dim):
+            """
+            Get the block dimension.
+
+            Get the block dimension.
+            """
+            return 0
+
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/decorators.py:166]: /bot accept 
this will be tested https://github.com/pyccel/pyccel-cuda/pull/66

**Comment by smazouz42** [comment]
bot mark as ready

File: pyccel/decorators.py [hunk 4]
```diff
@@ -130,6 +130,40 @@ def kernel(f):
...
+        def threadIdx(self, dim):
+            """
+            Get the thread index.
+
+            Get the thread index.
+            """
+            return self._thread_idx
+
+        def blockIdx(self, dim):
```
Comment (jalalium) [inline pyccel/decorators.py:151]: why not implement the logic for the dim argument from the start?

File: pyccel/decorators.py [hunk 5]
```diff
@@ -130,6 +130,40 @@ def kernel(f):
...
+            """
+            Get the block index.
+
+            Get the block index.
+            """
+            return self._block_idx
+
+        def blockDim(self, dim):
```
Comment (jalalium) [inline pyccel/decorators.py:159]: why not implement the logic for the dim argument from the start?

File: pyccel/decorators.py [hunk 6]
```diff
@@ -139,7 +173,19 @@ class KernelAccessor:
         def __init__(self, f):
             self._f = f
         def __getitem__(self, args):
-            return self._f
+            num_blocks, num_threads = args
```
Comment (jalalium) [inline pyccel/decorators.py:176]: Here too, num_blocks and num_threads are not necessarily one dimentional and can be up to three dimensions

**Review by jalalium — CHANGES_REQUESTED** [review]

File: pyccel/decorators.py [hunk 7]
```diff
@@ -130,6 +130,40 @@ def kernel(f):
...
+        """
+        Class representing the CUDA thread indexing.
+
+        Class representing the CUDA thread indexing.
+        """
+        def __init__(self, block_idx, thread_idx):
+            self._block_idx = block_idx
+            self._thread_idx = thread_idx
+
+        def threadIdx(self, dim):
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/decorators.py:143]: For the moment, you can't use tuples for num_blocks and num_threads. If you pass these values as tuples to the kernel call, you will receive a 'not implemented' error. Once this issue is resolved, I will implement the logic for the dim

File: pyccel/decorators.py (hunk 4 above)
Comment (smazouz42) [reply pyccel/decorators.py:151]: For the moment, you can't use tuples for num_blocks and num_threads. If you pass these values as tuples to the kernel call, you will receive a 'not implemented' error. Once this issue is resolved, I will implement the logic for the dim

File: pyccel/decorators.py (hunk 5 above)
Comment (smazouz42) [reply pyccel/decorators.py:159]: For the moment, you can't use tuples for num_blocks and num_threads. If you pass these values as tuples to the kernel call, you will receive a 'not implemented' error. Once this issue is resolved, I will implement the logic for the dim

File: pyccel/decorators.py (hunk 6 above)
Comment (smazouz42) [reply pyccel/decorators.py:176]: For the moment, you can't use tuples for num_blocks and num_threads. If you pass these values as tuples to the kernel call, you will receive a 'not implemented' error. Once this issue is resolved, I will implement the logic for the dim

File: CHANGELOG.md [hunk 8]
```diff
@@ -11,6 +11,7 @@ All notable changes to this project will be documented in this file.
 -   #42 : Add support for custom kernel in`cuda`.
 -   #42 : Add Cuda module to Pyccel. Add support for `cuda.synchronize` function.
 -   #41 : Add support for custom device in`cuda`.
+-   #41 : Improve kernel decorator.
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply CHANGELOG.md]: [738371a](https://github.com/pyccel/pyccel-cuda/pull/69/commits/738371a3dbf2fd73662800b2fb9716f5c24210bd)

**Review by jalalium — APPROVED** [review]

File: pyccel/decorators.py [hunk 9]
```diff
@@ -130,6 +130,40 @@ def kernel(f):
...
+            """
+            Get the thread index.
+
+            Get the thread index.
```
Comment (EmilyBourne) [inline pyccel/decorators.py:147]: Not sur why you repeat here? Either you don't want to follow the doc formats (which aren't enforced because this class is hidden inside a function) or you do and the parameter descriptions are missing

File: pyccel/decorators.py [hunk 10]
```diff
@@ -130,6 +130,40 @@ def kernel(f):
     KernelAccessor
         A class representing the kernel function.
     """
+    class CudaThreadIndexing:
+        """
+        Class representing the CUDA thread indexing.
+
+        Class representing the CUDA thread indexing.
+        """
```
Comment (EmilyBourne) [inline pyccel/decorators.py:138]: The classes inside this `kernel` function will be recreated at each kernel call. I think it would make more sense to put them somewhere in pyccel/cuda

File: pyccel/decorators.py [hunk 11]
```diff
@@ -139,7 +173,19 @@ class KernelAccessor:
...
+            num_blocks, num_threads = args
+            def internal_loop(*args, **kwargs):
+                """
+                The internal loop for kernel execution.
+
+                The internal loop for kernel execution.
+                """
+                for b in range(num_blocks):
+                    for t in range(num_threads):
+                        self._f.__globals__['cu'] = CudaThreadIndexing(b, t)
```
Comment (EmilyBourne) [inline pyccel/decorators.py:185]: What is the global `cu` ? I don't think this name will work when you try to retrieve the thread dimension and block dimension

File: CHANGELOG.md [hunk 12]
```diff
@@ -11,6 +11,7 @@ All notable changes to this project will be documented in this file.
 -   #42 : Add support for custom kernel in`cuda`.
 -   #42 : Add Cuda module to Pyccel. Add support for `cuda.synchronize` function.
 -   #41 : Add support for custom device in`cuda`.
+-   #69 : Improve kernel decorator.
```
Comment (EmilyBourne) [inline CHANGELOG.md:14]: This should go in the Changed section not the Added section, but I'm not actually sure that it needs to be here at all as the Change Log describes things that have changed since the previous release

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/decorators.py (hunk 9 above)
Comment (smazouz42) [reply pyccel/decorators.py]: The bot asks for documentation at https://github.com/pyccel/pyccel-cuda/pull/69/checks?check_run_id=27909189499, but I don't think it's necessary to include everything in the parameter and return sections

File: CHANGELOG.md (hunk 12 above)
Comment (smazouz42) [reply CHANGELOG.md]: [aa76f91](https://github.com/pyccel/pyccel-cuda/pull/69/commits/aa76f916848235c095e44be7c820d496cec000a5)

File: pyccel/decorators.py (hunk 11 above)
Comment (smazouz42) [reply pyccel/decorators.py]: [ec738b3](https://github.com/pyccel/pyccel-cuda/pull/69/commits/ec738b3b55f3f9ad23b227262f436f132477af43)

File: pyccel/decorators.py (hunk 9 above)
Comment (EmilyBourne) [reply pyccel/decorators.py]: Yes, pylint checks that there is documentation but nothing checks the format. So I would expect either:
```
            Get the thread index.
```
or
```
            Get the thread index.

            Get the thread index.

            Parameters
            -----------------
            dim : int
                   The dim argument that does something I guess
```

Comment (EmilyBourne) [reply pyccel/decorators.py]: You seem to have gone for something in the middle where you ignore the requirements that add useful information but keep the requirements that add useless information

File: pyccel/decorators.py (hunk 11 above)
Comment (EmilyBourne) [reply pyccel/decorators.py]: This change doesn't look right. I just want you to rename `cu` as the current implementation doesn't work. I don't think your new implementation works either.

Please add a test to make sure the threads/blocks/etc can actually be retrieved

Comment (EmilyBourne) [reply pyccel/decorators.py]: Do you understand what this line does and what `self._f.__globals__` is?

Comment (smazouz42) [reply pyccel/decorators.py]:
```py
# pylint: disable=missing-function-docstring, missing-module-docstring
from pyccel.decorators import kernel
from pyccel import cuda

@kernel
def say_hello(its_morning : bool):
    if(its_morning):
        print("Hello and Good morning")
    else:
        print("Hello and Good afternoon")

def f():
    its_morning = True
    say_hello[5,5](its_morning)
    cuda.synchronize()

if __name__ == '__main__':
    f()
```
he CUDA model gets overridden by the class CudaThreadIndexing, and we encounter the error:
AttributeError: 'CudaThreadIndexing' object has no attribute 'synchronize'
So using a dynamically assigned value to a dictionary key solves that problem, and the new implementation works just fine.

Comment (EmilyBourne) [reply pyccel/decorators.py]: I disagree I am pretty sure that this:
```python
@kernel
def say_hello_thread():
     print(cuda.threadIdx)

if __name__ == '__main__':
    say_hello_thread[5,5]()
```
gives different results on Python vs Cuda.

After this PR in Python it should print:
```
0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 4 4 4 4 4
```
I am pretty sure that your implementation will print:
```
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```

Comment (EmilyBourne) [reply pyccel/decorators.py]: > he CUDA model gets overridden by the class CudaThreadIndexing, and we encounter the error:
AttributeError: 'CudaThreadIndexing' object has no attribute 'synchronize'

In that case the class should not be `CudaThreadIndexing` but rather something like `CudaModuleInstance`

File: pyccel/decorators.py (hunk 9 above)
Comment (smazouz42) [reply pyccel/decorators.py]: 528099f is that better

File: pyccel/decorators.py (hunk 10 above)
Comment (smazouz42) [reply pyccel/decorators.py]: 528099f is that better

File: pyccel/decorators.py (hunk 11 above)
Comment (smazouz42) [reply pyccel/decorators.py]: i this that better 528099f

Comment (EmilyBourne) [reply pyccel/decorators.py]: A little but you should not need to do this:
https://github.com/pyccel/pyccel-cuda/commit/528099f81210532cc16f73fa1d79392b55fd4216#diff-fde82864b093f3b7f981abc6ec5374f7d4d5f4e695ed5d11363ee6eec8b2d2dcR14

synchronise should be inside the new class which is why I suggested renaming it `CudaModuleInstance`. With your current solution `cuda.threadIdx` will give the right result the first time it is called but the wrong result if it is called after the import

File: pyccel/decorators.py (hunk 9 above)
Comment (EmilyBourne) [reply pyccel/decorators.py]: Yes great thanks

File: pyccel/decorators.py (hunk 11 above)
Comment (smazouz42) [reply pyccel/decorators.py]: But cuda.threadIdx should be called only inside a kernel call

Comment (smazouz42) [reply pyccel/decorators.py]: I'm trying to separate the method of CUDA called from the kernel and not from the kernel.

Comment (smazouz42) [reply pyccel/decorators.py]: 57f977e

Comment (smazouz42) [reply pyccel/decorators.py]: @EmilyBourne

Comment (smazouz42) [reply pyccel/decorators.py]: or ```py
 for b in range(num_blocks):
                    for t in range(num_threads):
                        cu = CudaThreadIndexing(b, t)
                        cuda_globals = self._f.__globals__['cuda']
                        setattr(cuda_globals, 'threadIdx', cu.threadIdx)
                        setattr(cuda_globals, 'blockIdx', cu.blockIdx)
                        setattr(cuda_globals, 'blockDim', num_threads)
                        self._f(*args, **kwargs)
                        ```

File: pyccel/cuda/cuda_thread_indexing.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            - 0 for the x-dimension
+            - 1 for the y-dimension
+            - 2 for the z-dimension
+
+        Returns
+        -----------------
+        int
+            The size of the block in the specified dimension.
+        """
+        return 0
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/cuda/cuda_thread_indexing.py:87]: /bot accept
this will be tested https://github.com/pyccel/pyccel-cuda/pull/66

File: pyccel/cuda/cuda_thread_indexing.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+            - 1 for the y-dimension
+            - 2 for the z-dimension
+
+        Returns
+        -----------------
+        int
+            The index of the block in the specified dimension.
+        """
+        return self._block_idx
+
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/cuda/cuda_thread_indexing.py:67]: /bot accept
this will be tested https://github.com/pyccel/pyccel-cuda/pull/66

File: pyccel/cuda/cuda_thread_indexing.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+            - 1 for the y-dimension
+            - 2 for the z-dimension
+
+        Returns
+        -------
+        int
+            The index of the thread in the specified dimension of its block.
+        """
+        return self._thread_idx
+
```
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/cuda/cuda_thread_indexing.py:46]: /bot accept
this will be tested https://github.com/pyccel/pyccel-cuda/pull/66

File: pyccel/decorators.py (hunk 11 above)
Comment (EmilyBourne) [reply pyccel/decorators.py]: Ah! That sounds like a better idea :+1:
Good thinking :)

File: pyccel/cuda/cuda_thread_indexing.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/cuda/cuda_thread_indexing.py:46]: Can these be tested with a print? The analysis of the results will have to be done slightly differently but it should be possible.
E.g you could do one test for block_idx and one for thread_idx:
```python
from pyccel.decorators import kernel
from pyccel            import cuda

@kernel
def print_block():
    print(cuda.block_idx(0))

def f():
    print_block[5,5]()
    cuda.synchronize()

if __name__ == '__main__':
    f()
```
Instead of comparing strings, you would have to split the output into integers and check that all expected integers are present. E.g:
```python
python_block_idx = [int(word) for word in python_output]
for i in range(5):
    assert python_block_idx.count(i) == 5
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/cuda/cuda_thread_indexing.py:46]: Even if the translation is not yet available it should be possible to test this with a Python only test

File: tests/pyccel/scripts/kernel/device_test.py [hunk 16]
```diff
@@ -1,6 +1,6 @@
 # pylint: disable=missing-function-docstring, missing-module-docstring
 from pyccel.decorators import device, kernel
-from pyccel import cuda
+from pyccel            import cuda
 
 @device
```
Comment (EmilyBourne) [inline tests/pyccel/scripts/kernel/device_test.py:5]: Does `@device` also need access to the block_idx and thread_idx ?

File: pyccel/decorators.py [hunk 17]
```diff
@@ -139,7 +140,24 @@ class KernelAccessor:
...
+
+                The internal loop for kernel execution.
+                """
+                for b in range(num_blocks):
+                    for t in range(num_threads):
+                        cu = CudaThreadIndexing(b, t)
+                        if 'cuda' in self._f.__globals__:
+                            self._f.__globals__['cuda'].threadIdx = cu.threadIdx
+                            self._f.__globals__['cuda'].blockIdx = cu.blockIdx
+                            self._f.__globals__['cuda'].blockDim = cu.blockDim
```
Comment (EmilyBourne) [inline pyccel/decorators.py:156]: Please update the docs to make it clear that the user cannot use syntax such as:
```
from pyccel.decorators import kernel
from pyccel import cuda as py_cu

@kernel
def f():
    print(py_cu.block_idx)
```

Alternatively you may want to search for the correct key by iterating over `self._f.__globals__` and comparing the values to the `pyccel.cuda` module object

Comment (smazouz42) [reply pyccel/decorators.py:156]: I think the best thing is to inform the user about the valid and invalid syntax when we have the future in Pyccel. Regarding the documentation, I will add that in the PR when we have thread indexing. Or do you think it's better to add that now ??

File: tests/pyccel/scripts/kernel/device_test.py (hunk 16 above)
Comment (smazouz42) [reply tests/pyccel/scripts/kernel/device_test.py:5]: No, I import CUDA for cuda.synchronize()

File: pyccel/decorators.py (hunk 17 above)
Comment (EmilyBourne) [reply pyccel/decorators.py:156]: As long as you don't forget then it's ok to do it later. I am just worried that we will forget.

Pyccel can handle both syntaxes, it is just on the Python side that it is currently unhandled.

If you don't want to do this in this PR please create an issue

File: tests/pyccel/scripts/kernel/device_test.py (hunk 16 above)
Comment (EmilyBourne) [reply tests/pyccel/scripts/kernel/device_test.py:5]: Let me rephrase.

Should it be possible to access the block_idx and thread_idx from a device function? With the current implementation it is only possible to use these functions from a kernel function

File: pyccel/cuda/cuda_thread_indexing.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/cuda/cuda_thread_indexing.py:46]: Do we have a way to add tests like that to Pyccel ??

File: tests/pyccel/scripts/kernel/device_test.py (hunk 16 above)
Comment (smazouz42) [reply tests/pyccel/scripts/kernel/device_test.py:5]: Yes, and that is another thing to take into consideration

Comment (smazouz42) [reply tests/pyccel/scripts/kernel/device_test.py:5]: I'm going to open an issue about that

File: pyccel/decorators.py (hunk 17 above)
Comment (smazouz42) [reply pyccel/decorators.py:156]: [654afa3](https://github.com/pyccel/pyccel-cuda/pull/69/commits/654afa3c2938b406e1d54bf4eba26a407cccf297) Does that look good for now ??

Comment (EmilyBourne) [reply pyccel/decorators.py:156]: Yes great. Thanks

File: pyccel/cuda/cuda_thread_indexing.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/cuda/cuda_thread_indexing.py:46]: I'm not sure if you really need that, Emily, and if there's a way to test it without going beyond the scope of this PR. I don't see a need to test thread indexing without arrays, as testing with arrays should be easier and more relevant. and the PR for arrays depends on this one

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/cuda/cuda_thread_indexing.py:46]: >  if there's a way to test it without going beyond the scope of this PR

I am not sure how adding a test to ensure that the Python code gives the output that is expected after this PR is going beyond the scope of this PR?

> I don't see a need to test thread indexing without arrays

If thread indexing doesn't make sense without arrays then why do we need this PR?

> as testing with arrays should be easier

Are the 3 lines that I suggested complicated?

>  and the PR for arrays depends on this one

In that case we should be sure that this one works as intended. And we can't be sure without a test

[the remark this answers is not in the document]
Comment (smazouz42) [reply pyccel/cuda/cuda_thread_indexing.py:46]: Does that look better, Emily  ded14a7 ??

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/cuda/cuda_thread_indexing.py:46]: Yes. Great. Thanks :)

File: tests/pyccel/scripts/kernel/device_test.py (hunk 16 above)
Comment (EmilyBourne) [reply tests/pyccel/scripts/kernel/device_test.py:5]: ok. Please let me know when this is done

Comment (smazouz42) [reply tests/pyccel/scripts/kernel/device_test.py:5]: https://github.com/pyccel/pyccel-cuda/issues/70

Comment (EmilyBourne) [reply tests/pyccel/scripts/kernel/device_test.py:5]: #70

File: tests/pyccel/scripts/kernel/hello_kernel.py [hunk 18]
```diff
@@ -11,7 +11,7 @@ def say_hello(its_morning : bool):
 
 def f():
     its_morning = True
-    say_hello[1,1](its_morning)
+    say_hello[5,5](its_morning)
```
Comment (EmilyBourne) [inline tests/pyccel/scripts/kernel/hello_kernel.py:14]: Please don't change this.
It is not impossible that the output comes out as:
`Hello and Hello and Good morningGood morning`
or something horrible like this. Having a 1,1 test makes this impossible

Comment (smazouz42) [reply tests/pyccel/scripts/kernel/hello_kernel.py]: https://github.com/pyccel/pyccel-cuda/pull/69/commits/ef09298c9793b6fc1dda82edb0b53a07c97b3fd3

**Review by EmilyBourne — APPROVED** [review]
Looks good to me. This should make it much easier to test our array devs!
